_pages/clients/rust.md

```diff
@@ -63,200 +61,200 @@ If you are running [Algorithmia Enterprise](/enterprise), you can specify the AP
 
-### Downloading Data from a Data Collection
+## Working with Algorithmia Data Sources
 
-Next check if the file that you just uploaded to data collections exists, and try downloading it to a (new) local file:
+For some algorithms, passing input to the algorithm at request time is sufficient, while others might have larger data requirements or need to preserve state between calls. Application developers can use Algorithmia's [Hosted Data](/developers/data/hosted) to store data as text, JSON, or binary, and access it via the Algorithmia [Data API](/developers/api/#data).
 
-{% highlight rust %}
-// Download the file
-if client.file(text_file).exists().unwrap() {
-	let mut file_reader = client.file(text_file).get().unwrap();
-	let mut localfile = File::create("/path/to/save/localfile.png").unwrap();
-	std::io::copy(&mut file_reader, &mut localfile);
-}
-{% endhighlight %}
+The Data API defines [connectors](/developers/api/#connectors) to a variety of storage providers, including Algorithmia [Hosted Data](/developers/data/hosted), Amazon S3, Google Cloud Storage, Azure Storage Blobs, and Dropbox. After creating a connection in Data Sources, you can use the API to create, update, and delete directories and files and manage permissions across providers by making use of [Data URIs](/developers/api/#data-uris) in your code.
 
-This copies the file from your data collection and saves it as a file on your local machine, storing the filename in the variable `localfile`.
+In this example, we'll upload an image to Algorithmia's [Hosted Data](/developers/data/hosted) storage provider, and use the [dlib/FaceDetection](https://algorithmia.com/algorithms/dlib/FaceDetection) algorithm to detect any faces in the image. The algorithm will create a new copy of the image with bounding boxes drawn around the detected faces, and then return a JSON object with details about the dimensions of the bounding boxes and a URI where you can download the resulting image.
 
-Alternately, if you just need the text content of the file to be stored in a variable, you can retrieve the remote file's content without saving the actual file:
+### Create a Data Collection
+
+The documentation for "Face Detection" says that it takes a URL or a Data URI of the image to be processed, and a Data URI where the algorithm can store the result. We'll create a directory to host the input image and set its [permissions](/developers/api/#update-collection-acl) to make it publicly accessible by creating a DataAcl with `DataAcl::from(ReadAcl::Public)` and passing it to the `create` method.
 
 {% highlight rust %}
-// Download contents of file as a string
-if client.file(text_file).exists().unwrap() {
-	let mut text_reader = client.file(text_file).get().unwrap();
-	let mut jack_london_text = String::new();
-	let _ = text_reader.read_to_string(&mut jack_london_text);
+let img_directory = client.dir("data://YOUR_USERNAME/img_directory");
+if img_directory.exists().unwrap() == false {
```

Comment:
nit: testing for `== false` followed by an `else` clause is less natural to read than:

```
if img_directory.exists()? {
   println!("img_directory exists");
} else { 
    img_directory.create(DataAcl::from(ReadAcl::Public))?;
}
```

(no strong opinion here on `?` vs `unwrap()` in example code)

Reply:
Looks good, updated in two places.